src/utils/auth.ts

```diff
@@ -0,0 +1,225 @@
+import {
+  loginRequest,
+  tokenRequest,
+  silentRequest,
+  msalConfig,
+} from '../config/authConfig';
+
+class AuthProvider {
+  private publicClient: PublicClientApplication;
+
+  private _account: AccountInfo | null = null;
+  private _token: AuthenticationResult | null = null;
+
+  private loginPromise: Promise<void> | null = null;
+  private initPromise: Promise<void[]>;
+
+  get isAuthenticated() {
+    return this._token && this._token.expiresOn > new Date();
+  }
+
+  get account() {
+    return this._account;
+  }
+
+  get token() {
+    return this._token;
+  }
+
+  constructor(config: Configuration) {
+    this.publicClient = new PublicClientApplication(config);
+
+    console.log('starting auth provider');
+
+    const initPromises: Promise<void>[] = [];
+
+    initPromises.push(
+      (async (publicClient: PublicClientApplication) => {
+        try {
+          const response = await publicClient.handleRedirectPromise();
+          this.handleLoginResponse(response, null);
+        } catch (err) {
+          this.handleLoginResponse(null, err);
+        }
+      })(this.publicClient)
+    );
+
+    const account = this.getActiveAccount();
+
+    if (account) {
+      console.log('found cached account', account);
+      this._account = account;
+    }
+
+    initPromises.push(
+      (async () => {
+        if (this.loginPromise) await this.loginPromise;
+        this.loginPromise = this.login(loginRequest);
+        await this.loginPromise;
+      })()
+    );
+
+    this.initPromise = Promise.all(initPromises);
+  }
+
+  private handleLoginResponse(
+    response: AuthenticationResult | null,
+    err: Error | null
+  ) {
+    if (err) {
+      console.error('login error', err);
+      return;
+    }
+
+    console.log('login result', response);
+
+    const account = this.getActiveAccount();
+
+    if (response) {
+      console.log('login success, has response', account);
+      this._account = account;
+      this._token = response;
+    } else {
+      if (account) {
+        console.log('has accounts', account);
+        this._account = account;
+      }
+    }
+  }
+
+  private handleApiTokenResponse(
+    response: AuthenticationResult | null,
+    err: Error | null
+  ) {
+    if (err) {
+      console.error('api token error', err);
+      return;
+    }
+
+    console.log('api token result', response);
+
+    if (response) {
+      if (!response.accessToken) {
+        console.error(
+          'api token failed, got response but no accessToken field'
+        );
+      }
+    } else {
+      console.error('api token failed, null response but no error');
+    }
+  }
+
+  private getActiveAccount(): AccountInfo | null {
+    const allAccounts = this.publicClient.getAllAccounts();
+
+    if (allAccounts?.length) return allAccounts[0];
+
+    return null;
+  }
+
+  private async login(loginRequest: RedirectRequest) {
+    if (this.isAuthenticated) {
```

Comment:
elaborate 😄

Reply:
Sorry for the late reply, procrastination became vacationing 🙈.

It's just my personal preference, but I like to be able to distinguish between fields and computed values. The reason for this is because it _can_ be more expensive to resolve a computed value than a field value. As such, a caller might unknowingly be invoking expensive operations without having considered the underlying cost; all due to the incorrect assumption that they are resolving a field value - as both operations have the same syntax.